File: brand-service/src/main/java/com/example/brandservice/model/Promotion.java
package com.example.brandservice.model;

import com.example.brandservice.converter.GameListConverter;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Data;
import lombok.experimental.FieldDefaults;

import java.time.LocalDateTime;
import java.util.Date;
import java.util.List;

@Entity
@Data
@Table(name = "promotions")
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Promotion {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    String id;
    String name;
    String description;
    String imageUrl;
    Date startDate;
    Date endDate;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "brandId", referencedColumnName = "id")
    Brand brand;

    Double budget;
    Double remainingBudget;

   @Enumerated(EnumType.STRING)
    PromotionStatus status;

    @JsonManagedReference
    @OneToMany(mappedBy = "promotion", fetch = FetchType.LAZY)
    List<Voucher> vouchers;

    @Convert(converter = GameListConverter.class)
    @Column(columnDefinition = "TEXT")
    List<String> games;

    @Column(name = "createdAt")
    LocalDateTime createAt;
    @Column(name = "updatedAt")
    LocalDateTime updateAt;

    public enum PromotionStatus {
        ACTIVE,
        INACTIVE,
        EXPIRED,
        PAID
    }
}

File: game-service/models/Promotion.js
const mongoose = require("mongoose");
const { v4: uuidv4 } = require("uuid");

const PromotionSchema = new mongoose.Schema(
	{
		_id: { type: String, default: uuidv4 }, // UUID
		name: { type: String, required: true }, // Tên khuyến mãi
		description: { type: String }, // Mô tả khuyến mãi
		imageUrl: { type: String }, // Hình ảnh khuyến mãi
		startDate: { type: Date, required: true }, // Thời gian bắt đầu
		endDate: { type: Date, required: true }, // Thời gian kết thúc
		brandId: { type: String, required: true }, // Thương hiệu liên kết
		budget: { type: Number, default: 0 }, // Ngân sách dành cho khuyến mãi
		remainingBudget: { type: Number }, // Ngân sách còn lại
		status: {
			type: String,
			enum: ["active", "inactive", "completed"],
			default: "active",
		}, // Trạng thái khuyến mãi
		vouchers: [{ type: String }], // Danh sách voucher
		games: [{ type: String, ref: "Game" }], // Danh sách trò chơi
	},
	{ timestamps: true }
);

PromotionSchema.set("toJSON", {
	virtuals: true,
});

module.exports = mongoose.model("Promotion", PromotionSchema);
